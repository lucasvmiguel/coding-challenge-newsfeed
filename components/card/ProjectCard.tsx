import Link from 'next/link'
import styled from 'styled-components'

import Card, { Icon, Columns, ColumnLeft, ColumnRight } from './Card'
import Markdown from '../Markdown'

type User = {
  id: number;
  name: string;
  avatar_url: string;
}

export type Project = {
  id: number;
  name: string;
  description: string;
  icon_url: string;
  users: User[];
}

type Props = {
  project: Project;
}

export default function ProjectCard({ project }: Props) {
  return (
    <Card>
      <Columns>
        <ColumnLeft>
          <Icon src={project.icon_url} />
        </ColumnLeft>
        <ColumnRight>
          <h2 data-test="project-card-name">{project.name}</h2>
          <Markdown>{project.description}</Markdown>
          {!!project.users.length && (
            <>
              <h3>Participants:</h3>
              {project.users.map(u => (
                <Participant key={u.id} user={u} />
              ))}
            </>
          )}
        </ColumnRight>
      </Columns>
    </Card>
  )
}

function Participant({ user }: { user: User }) {
  return (
    <ParticipantContainer>
      <ParticipantColumnLeft>
        <ParticipantAvatar src={user.avatar_url} />
      </ParticipantColumnLeft>
      <ParticipantColumnRight>
        <Link href={`/users/${user.id}`}>
          {user.name}
        </Link>
      </ParticipantColumnRight>
    </ParticipantContainer>
  )
}

const ParticipantAvatar = styled.img`
  border-radius: 3px;
  background-color: rgba(0, 0, 0, 0.1);
`

const ParticipantContainer = styled.div`
  display: flex;
  flex-direction: row;
  margin-bottom: 1rem;
`

const ParticipantColumnLeft = styled.div`
  flex-basis: 2rem;
  flex-shrink: 0;
  flex-grow: 0;
  margin-right: 1rem;
`

const ParticipantColumnRight = styled.div`
  flex-basis: 3rem;
  flex-grow: 1;
  display: flex;
  flex-direction: column;
  justify-content: center;
`
