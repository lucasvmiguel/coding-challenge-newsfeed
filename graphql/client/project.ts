import { gql } from '@apollo/client'

export const PROJECT_QUERY = gql`
  query project($id: Int!) {
    project(id: $id) {
      id
      name
      description
      icon_url
      users {
        id
        name
        avatar_url
      }
    }
  }
`
