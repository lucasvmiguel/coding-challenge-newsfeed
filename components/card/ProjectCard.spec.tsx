import renderer from 'react-test-renderer'

import { project } from '../../test/mocks/project'

import ProjectCard from './ProjectCard'

it('renders ProjectCard component correctly - snapshot', () => {
  const tree = renderer
    .create(<ProjectCard project={project} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})