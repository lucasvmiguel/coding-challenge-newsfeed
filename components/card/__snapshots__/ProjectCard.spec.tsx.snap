// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders ProjectCard component correctly - snapshot 1`] = `
<div
  className="sc-bdfBQB eDCjUP"
>
  <div
    className="sc-dlfnuX iOZAXF"
  >
    <div
      className="sc-hKgJUU jkblDv"
    >
      <img
        className="sc-gsTEea iRQMQw"
        src="icon.svg"
      />
    </div>
    <div
      className="sc-eCstlR evEecB"
    >
      <h2
        data-test="project-card-name"
      >
        name testing
      </h2>
      <div
        className="sc-jSgvzq"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<p>desc testing</p>
",
          }
        }
      />
      <h3>
        Participants:
      </h3>
      <div
        className="sc-iBPTik ddGxJq"
      >
        <div
          className="sc-fubCzh kGMoIv"
        >
          <img
            className="sc-gKseQn gmHRui"
            src="icon2.svg"
          />
        </div>
        <div
          className="sc-pGacB dXoMBx"
        >
          <a
            href="/users/2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            user testing
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
